src/main/kotlin/com/nikitanazarov/coroutinestacks/CoroutineTraceForestBuilder.kt

```diff
@@ -0,0 +1,258 @@
+package com.nikitanazarov.coroutinestacks
+
+import com.intellij.debugger.engine.SuspendContextImpl
+import com.intellij.openapi.application.ApplicationManager
+import com.intellij.ui.JBColor
+import com.intellij.ui.components.JBList
+import com.intellij.ui.components.JBScrollPane
+import com.intellij.ui.util.preferredWidth
+import com.intellij.util.ui.JBUI
+import com.intellij.xdebugger.frame.XExecutionStack
+import com.intellij.xdebugger.frame.XStackFrame
+import com.nikitanazarov.coroutinestacks.ui.ContainerWithEdges
+import com.nikitanazarov.coroutinestacks.ui.ForestLayout
+import com.nikitanazarov.coroutinestacks.ui.Separator
+import com.sun.jdi.Location
+import org.jetbrains.kotlin.analysis.decompiler.stub.file.ClsClassFinder
+import org.jetbrains.kotlin.idea.debugger.coroutine.data.CoroutineInfoData
+import org.jetbrains.kotlin.idea.debugger.coroutine.data.CoroutineStackFrameItem
+import org.jetbrains.kotlin.idea.debugger.coroutine.data.State
+import org.jetbrains.kotlin.idea.debugger.coroutine.util.CoroutineFrameBuilder
+import java.awt.*
+import java.awt.event.MouseAdapter
+import java.awt.event.MouseEvent
+import java.util.*
+import javax.swing.border.Border
+import javax.swing.border.LineBorder
+
+data class Node(
+    val stackFrameItem: CoroutineStackFrameItem? = null,
+    var num: Int = 0, // Represents how many coroutines have this frame in their stack trace
+    val children: MutableMap<Location, Node> = mutableMapOf(),
+    var coroutinesActive: String = ""
+)
+
+data class CoroutineTrace(val stackFrameItems: MutableList<CoroutineStackFrameItem?>, val header: String, val coroutinesActiveLabel: String)
+
+private fun CoroutineInfoData.render(): String =
```

Comment:
Let's put this function in the bottom of the file, since it is private anyway and doesn't expose any interesting api, and is used only for internal implementation of this file